Recursion (#1)
Ada-Developers-Academy/core-interlude: recursion/intro-to-recursion.md

# Intro to Recursion

## Learning Goals

- Define recursion

## Introduction

> A powerful problem-solving technique in computer science is to use recursion. It takes advantage of operations that computers excel at.

Reviewer: bold claims in this passage!
- a powerful technique (how?)
- is to use recursion (but what is it?) 
- operations that computers excel at (like???)

Instead, how about we keep it light with a bit of a ridiculous "definition," but one that can still get the reader thinking before we dive in further?

```suggestion
Life is full of problems. Big problems. But sometimes big problems are based on a smaller version of the same problem. And that smaller problem is itself based on a yet smaller version of the same problem. And that yet smaller problem is itself based on an even smaller version of the same problem, all the way down until we reach a size of the problem that isn't so big, and that we know how to solve.

Such problems are said to be recursive.
```

> However, recursion happens around us all the time! Observing and naming recursion is a great way to begin thinking about recursion.

Reviewer: basically the same sentence, but reworked a bit to flow better from the previous change.
```suggestion
But recursion applies to more than just problems. Recursion happens around us all the time. Observing and naming recursion in the real world is a great way to begin thinking about recursion in our programs!
```

## Vocabulary and Synonyms

| Vocab     | Definition                                           | How to Use in a Sentence                                                                                     |
| --------- | ---------------------------------------------------- | ------------------------------------------------------------------------------------------------------------ |
| Recursion | The process of defining something in terms of itself | "The acronym GNU means "GNU's Not Unix." This is a recursive acronym, because its meaning refers to itself." |

## Recursion

Recursion is what happens when something's definition refers to itself. This could mean that something is constructed using itself, or that its construction is based on itself.

> A recursive algorithm solves a problem by using the solving a smaller version of the same problem.

Reviewer: ```suggestion
A recursive algorithm solves a problem by solving a smaller version of the same problem.
```

## Examples from Mathematics

The [Sierpiński triangle](https://en.wikipedia.org/wiki/Sierpi%C5%84ski_triangle) is a triangle which contains smaller triangles. These smaller triangles contain other smaller triangles... which contain other smaller triangles. This pattern goes on infinitely.

The Sierpiński triangle is recursive because it is made up of other Sierpiński triangles.

The [Koch snowflake](https://en.wikipedia.org/wiki/Koch_snowflake) (also known as Koch curve) is a shape made using a recursive algorithm.

> A Koch snowflake is created by starting with an equilateral triangle, and following this algorithm for each line segment in the shape:

Reviewer: help direct the reader's attention to the algorithm that comes after the text.
```suggestion
A Koch snowflake is created by starting with an equilateral triangle, and applying the following algorithm for each line segment in the shape:
```

1. Divide the line segment into three segments of equal length
1. Draw an equilateral triangle, whose base is the middle segment from step one
1. Remove the line segment that is the base of the triangle from step two

> 

Reviewer: the fibonacci sequence is mentioned below, so this seems like a good time to give an overview.
```suggestion

The [Fibonacci sequence](https://en.wikipedia.org/wiki/Fibonacci_number) is a sequence of numbers where each next number is defined in terms of previous numbers in the sequence. As the sequence grows, the ratios between consecutive terms approaches the golden ratio **𝛗**, which is often considered a marker of natural beauty.

```

## Examples from Nature

- Romanesco broccoli, where each bud is made up of smaller buds
- Sourdough bread that was made from leftover sourdough
- Branching behavior in certain plants and algae, where a branch branches off another branch

> We can observe many fractals and fibonacci sequences, both of which are recursive in nature.

Reviewer: give a light explanation of fractals
soften claims about fibonacci numbers (golden ratio) in nature (turns out it's overblown!)
```suggestion
We can observe many fractal patterns—patterns which have internal similarity—in nature. They can often be described recursively.

Many claim to find the golden ratio (related to the recursive Fibonacci sequence) hidden within numerous natural structures.
```

- Sunflower seed heads
- Pinecone seeds
- Flower petals

## Examples from Art

- Matryoshka dolls
  - Wooden dolls that contain other matryoshka dolls within itself
- Droste effect
  - An effect where an image contains the image itself
- Infinity mirrors

## Check for Understanding

<!-- Question 1 -->
<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: sE7nhe
* title: Thinking Recursively
##### !question

> Which of the following are obvious examples of recursion?

Reviewer: what if it's not obvious to the reader?
```suggestion
Which of the following are examples of recursion?
```

##### !end-question
##### !options

> * A triangle which contains other triangles

Reviewer: I think this is the only example that benefitted from "obvious" being called out in the question. So maybe make this more clearly _not_ recursive?
```suggestion
* A triangle which contains another triangle
```

* A Sierpiński triangle made up of other Sierpiński triangles
* Calculating the fibonacci sequence, in which the next number in the sequence is the sum of the last two numbers in the sequence
* A row of similar jars and containers on a shelf
* A painted self-portrait

##### !end-options
##### !answer

> * A Sierpiński triangle made up of other Sierpiński triangles
> * Calculating the fibonacci sequence, in which the next number in the sequence is the sum of the last two numbers in the sequence

Reviewer: expand explanations and make the format a little nicer
```suggestion
A Sierpiński triangle made up of other Sierpiński triangles is defined by referring to itself.

<br />

Calculating the fibonacci sequence follows a recursive algorithm. Its solution depends on the solutions of smaller versions of the same problem.

<br />

Putting a triangle in another triangle doesn't necessarily make it recursive. If that inner triangle also had an inner triangle, and _that_ inner triangle also had an inner triangle, and so on, then it could be recursive.

<br />

Similar jars aren't defined in terms of each other, so would not be recursive. They just happen to be similar.

<br />

A painted self portrait where the artist was painting a self portrait of themselves painting a self portrait, and so on, could be recursive, as in the Droste effect. But a typical self portrait would not be recursive.
```

##### !end-answer
##### !explanation

A Sierpiński triangle made up of other Sierpiński triangles is defined by referring to itself.

Calculating the fibonacci sequence follows a recursive algorithm. Its solution depends on the solutions of smaller versions of the same problem.

##### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question Takeaway -->
<!-- prettier-ignore-start -->
### !challenge
* type: paragraph
* id: xFPTHi

> * title: Thinking Recursively

Reviewer: match the lesson name
```suggestion
* title: Intro to Recursion
```

We can observe many fractal patterns—patterns which have internal similarity—in nature. They can often be described recursively.

Many claim to find the golden ratio (related to the recursive Fibonacci sequence) hidden within numerous natural structures.

- Sunflower seed heads
- Pinecone seeds
- Flower petals

## Examples from Art

- Matryoshka dolls
  - Wooden dolls that contain other matryoshka dolls within itself
- Droste effect
  - An effect where an image contains the image itself
- Infinity mirrors

## Check for Understanding

<!-- Question 1 -->
<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: sE7nhe
* title: Thinking Recursively
##### !question

Which of the following are examples of recursion?

##### !end-question
##### !options

* A triangle which contains another triangle
* A Sierpiński triangle made up of other Sierpiński triangles
* Calculating the fibonacci sequence, in which the next number in the sequence is the sum of the last two numbers in the sequence
* A row of similar jars and containers on a shelf
* A painted self-portrait

##### !end-options
##### !answer

> A Sierpiński triangle made up of other Sierpiński triangles is defined by referring to itself.
> 
> <br />
> 
> Calculating the fibonacci sequence follows a recursive algorithm. Its solution depends on the solutions of smaller versions of the same problem.
> 
> <br />
> 
> Putting a triangle in another triangle doesn't necessarily make it recursive. If that inner triangle also had an inner triangle, and _that_ inner triangle also had an inner triangle, and so on, then it could be recursive.
> 
> <br />
> 
> Similar jars aren't defined in terms of each other, so would not be recursive. They just happen to be similar.
> 
> <br />
> 
> A painted self portrait where the artist was painting a self portrait of themselves painting a self portrait, and so on, could be recursive, as in the Droste effect. But a typical self portrait would not be recursive.

Reviewer: maybe a bad merge?
```suggestion
* A Sierpiński triangle made up of other Sierpiński triangles
* Calculating the fibonacci sequence, in which the next number in the sequence is the sum of the last two numbers in the sequence
```

##### !end-answer
##### !explanation

> A Sierpiński triangle made up of other Sierpiński triangles is defined by referring to itself.
> 
> Calculating the fibonacci sequence follows a recursive algorithm. Its solution depends on the solutions of smaller versions of the same problem.

Reviewer: same merge issue?
```suggestion
A Sierpiński triangle made up of other Sierpiński triangles is defined by referring to itself.

<br />

Calculating the fibonacci sequence follows a recursive algorithm. Its solution depends on the solutions of smaller versions of the same problem.

<br />

Putting a triangle in another triangle doesn't necessarily make it recursive. If that inner triangle also had an inner triangle, and _that_ inner triangle also had an inner triangle, and so on, then it could be recursive.

<br />

Similar jars aren't defined in terms of each other, so would not be recursive. They just happen to be similar.

<br />

A painted self portrait where the artist was painting a self portrait of themselves painting a self portrait, and so on, could be recursive, as in the Droste effect. But a typical self portrait would not be recursive.
```
